### PR TITLE
Introduce llama-run

### PR DESCRIPTION
**Pull request by ericcurtin**
It's like simple-chat but it uses smart pointers to avoid manual
memory cleanups. Less memory leaks in the code now. Avoid printing
multiple dots. Split code into smaller functions. Uses no exception
handling.

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [ ] Low
  - [x] Medium
  - [ ] High

### CODE REVIEW

**Comment by ericcurtin** [comment]
Some of these builds seem hardcoded to c++11, when we use a feature from c++14.

Any reason we aren't using say c++17

Any reasonable platform should be up-to date with c++17 I think

**Comment by ericcurtin** [comment]
Converted to C++11 only

**Comment by ericcurtin** [comment]
@slaren @ggerganov PTAL, I'm hoping to add other features to this example such as, read prompt from '-p' arg and read prompt from stdin

**Comment by ericcurtin** [comment]
Cool sounds good @slaren, mind if I call the new example ramalama-core ?

**Comment by slaren** [comment]
I would prefer a more neutral name, but I will defer to @ggerganov about that.

Btw, it would be good to have smart pointer types in `llama.h` (guarded by an `#ifdef __cplusplus`) similar to the ones in https://github.com/ggerganov/llama.cpp/blob/master/ggml/include/ggml-cpp.h

**Comment by ericcurtin** [comment]
> I would prefer a more neutral name, but I will defer to @ggerganov about that.
> 
> Btw, it would be good to have smart pointer types in `llama.h` (guarded by an `#ifdef __cplusplus`) similar to the ones in https://github.com/ggerganov/llama.cpp/blob/master/ggml/include/ggml-cpp.h

It's a pity, RamaLama doesn't have the smart pointers everywhere approach (except in exception circumstances). It takes a while to get used to, but once you get the hang of it, it's not too bad.

**Comment by ericcurtin** [comment]
You can do things like:

```
git diff | llama-ramalama-core -m some-model -ngl -p "Write a git commit message for this change:"
```

with this example.

**Comment by ggerganov** [comment]
@ericcurtin Do you plan to develop this into a more advanced example/application in the future? With the existing functionality, I am not sure it has enough value to justify adding it to the project and maintaining it.

**Comment by ericcurtin** [comment]
@ggerganov I plan to continue adding more features.

**Comment by ericcurtin** [comment]
> I would prefer a more neutral name, but I will defer to @ggerganov about that.
> 
> Btw, it would be good to have smart pointer types in `llama.h` (guarded by an `#ifdef __cplusplus`) similar to the ones in https://github.com/ggerganov/llama.cpp/blob/master/ggml/include/ggml-cpp.h

Btw, did you mean as part of this PR or in general?

**Comment by ggerganov** [comment]
No need to be in this PR, just in general. If the example will continue to be developed, then it would make sense to merge it now.

**Comment by slaren** [comment]
Any name based on the functionality of the example is fine. "ramalama-core" doesn't mean anything to me, or to most users I would expect. The smart pointers can be added in this PR or a separate one, it doesn't matter to me. I mentioned it because since you are already using them, it may be worth to spend the time to make proper ones using structs instead of function pointers (which are very annoying to use due to having to initialize it with the pointer at every time, and probably also less efficient since it stores an additional pointer in the `unique_ptr`).

**Comment by ericcurtin** [comment]
llama-inferencer? I really don't care too much about the name, just want to agree on one to unblock this PR.

@rhatdan @slp any suggestions on names? I plan on using this as the main program during "ramalama run" but I'm happy for anyone to use it or make changes to it to suit their needs. It's like a drastically simplified version of llama-cli, with one additional feature, read from stdin.

But it does seem stabler and less error prone than llama-cli also. And the verbose info is all cleaned up to only spit out errors. It was based on llama-simple-chat initially.

**Comment by slaren** [comment]
Maybe something like `llama-chat`. I mentioned before that I think it would be good to have a example focused on chat only, that does that very well, that in time could replace the current `llama-cli` as the main program of llama.cpp, which at this is point is basically unmaintainable and should be retired.

**Comment by ericcurtin** [comment]
It's all tempting to call this something like llama-run and using a kinda RamaLama, LocalAI, Ollama type interface to interact with models. Kinda like daemonless Ollama:

llama-run file//somedir/somefile.gguf

**Comment by ericcurtin** [comment]
We could even possibly add https://, http://, ollama://,  hf:// since they all are just a http pull in the end

**Comment by ericcurtin** [comment]
That might be implemented as llama-pull that llama-run can fork/exec (or they share a common library)

**Comment by ericcurtin** [comment]
I will be afk for 3 weeks, so expect inactivity in this PR. I did the rename in case we want to merge as is and not leave this go stale.

Although the syntax will completely change to:

llama-run [file://]somedir/somefile.gguf

file will be optional, but will set up the possibility of adding the pullers discussed above.

**Comment by ericcurtin** [comment]
This drives the compiler crazy FWIW:

```
diff --git a/include/llama.h b/include/llama.h
index 5e742642..c3285da3 100644
--- a/include/llama.h
+++ b/include/llama.h
@@ -537,6 +537,13 @@ extern "C" {
                          int32_t   il_start,
                          int32_t   il_end);

+#ifdef __cplusplus
+    // Smart pointers
+    typedef std::unique_ptr<llama_model, decltype(&llama_free_model)> llama_model_ptr;
+    typedef std::unique_ptr<llama_context, decltype(&llama_free)> llama_context_ptr;
+    typedef std::unique_ptr<llama_sampler, decltype(&llama_sampler_free)> llama_sampler;
+#endif
+
     //
     // KV cache
     //
```

seems like it only wants to build C code in that file or something.

**Comment by slaren** [comment]
You are probably doing something wrong. Put it at the end of the file, include `<memory>`, and rename the `llama_sampler` ptr to `llama_sampler_ptr`. You should also structs (function objects) as the deleter, see `ggml-cpp.h` for an example.

**Comment by ericcurtin** [comment]
Added the smart pointers typedef stuff @slaren

File: include/llama-cpp.h
Comment (slaren) [inline include/llama-cpp.h:24]: The rest look good, but this one does not belong here.

Comment (ericcurtin) [reply include/llama-cpp.h:24]: I also need to use more parameters as references, I notice I still pass raw pointers in places where it could be a reference to a smart pointer

Comment (ericcurtin) [reply include/llama-cpp.h]: I removed char_array_ptr

**Review by slaren — APPROVED** [review]
Looks like a good start for an improved chat example.

`llama-cpp.h` also needs to be added to the list of public headers in `CMakeLists.txt`.

**Comment by ericcurtin** [comment]
Added as public header

**Review by ggerganov — APPROVED** [review]

**Comment by ericcurtin** [comment]
Happy to merge this @slaren @ggerganov 

Although there will be breaking changes in a follow on PR, want to change the syntax to this (ramalama/ollama style):

```
Description:
  Runs a llm

Usage:
  llama-run [options] MODEL [PROMPT]

Options:
  -c, --context-size <value>
      Context size (default: 2048)
  -n, --ngl <value>
      Number of GPU layers (default: 0)
  -h, --help
      Show help message

Examples:
  llama-run ~/.local/share/ramalama/models/ollama/smollm\:135m
  llama-run --ngl 99 ~/.local/share/ramalama/models/ollama/smollm\:135m
  llama-run --ngl 99 ~/.local/share/ramalama/models/ollama/smollm\:135m Hello World
```

Eventually want to add optional prefixs for the model also (like file://, hf://, maybe we implement http://, https://, ollama:// eventually also, its not that hard), but for no prefix we will just assume the string is a file path I guess.

**Comment by slaren** [comment]
Sounds good. Btw, to support dynamic loading of backends (#10469), you should add a call to `ggml_backend_load_all` at the beginning.

**Pull request merged**
